### PR TITLE
DynSolEvent & ResolveSolEvent

### PR DESCRIPTION
**Pull request by prestwich**
based on #271 

## Motivation

- Make clear checked/unchecked API for `Log` to allow safe usage as well as instantiating invalid logs
- Add `DynSolEvent`
- Add `ResolveSolEvent` to allow future reuse by non-JSON (e.g. `impl ResolveSolEvent for syn_solidity::ItemEvent` would be cool)

## PR Checklist

- [x] Added Tests
- [x] Added Documentation
- [ ] Breaking changes

### CODE REVIEW

**Review by Evalir — COMMENTED** [review]
ACK—quite like the checked/unchecked API myself.

**Comment by prestwich** [comment]
should the unchecked API be `unsafe`?  e.g. `unsafe fn new_unchecked` ?

**Comment by DaniPopes** [comment]
My stance on that is if an invalid value causes either UB, or panics everywhere to the point it's unusable. If it only causes a panic in small sections or not at all, then it should not be marked as unsafe. I don't think this is the case here

**Comment by prestwich** [comment]
cool. we'll leave them as safe then

File: crates/dyn-abi/src/event.rs [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+        let log = Log::new_unchecked(
+            vec![
+                t0,
+                b256!("0000000000000000000000000000000000000000000000000000000000012321"),
+            ],
+            bytes!(
+                "
+			0000000000000000000000000000000000000000000000000000000000012345
+			0000000000000000000000000000000000000000000000000000000000054321
+			"
```
Comment (gakonst) [inline crates/dyn-abi/src/event.rs:184]: nit de indent

File: crates/dyn-abi/src/event.rs [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+            body: DynSolType::Tuple(vec![DynSolType::Tuple(vec![
+                DynSolType::Address,
+                DynSolType::Address,
+            ])]),
+        };
+
+        let decoded = event.decode_log(&log, true).unwrap();
+        assert_eq!(
+            decoded.indexed,
+            vec![DynSolValue::Address(address!(
```
Comment (gakonst) [inline crates/dyn-abi/src/event.rs:199]: maybe eventually we add `sol_address!` as extra sugar, but might be overkill

File: crates/dyn-abi/src/event.rs [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+            match topics.size_hint() {
+                (n, Some(m)) if n == m && n != num_topics => {
+                    return Err(Error::TopicLengthMismatch {
+                        expected: num_topics,
+                        actual: n,
+                    })
+                }
+                _ => {}
+            }
+        }
```
Comment (gakonst) [inline crates/dyn-abi/src/event.rs:69]: 👍

**Review by gakonst — APPROVED** [review]

File: crates/dyn-abi/src/ext/event.rs [hunk 4]
```diff
@@ -1,15 +1,50 @@
...
+            return Err(Error::TopicLengthMismatch {
+                expected: 4,
+                actual: num_topics,
+            })
+        }
+
+        Ok(DynSolEvent::new_unchecked(
+            topic_0,
+            indexed,
+            DynSolType::Tuple(body),
```
[the remark this answers is not in the document]
Comment (prestwich) [reply crates/dyn-abi/src/ext/event.rs:43]: well, it's really just a sequence of types. but Tuple is the way to represent that. could be some cleanup here tho

File: crates/dyn-abi/src/event.rs (hunk 3 above)
Comment (prestwich) [reply crates/dyn-abi/src/event.rs:69]: this is @DaniPopes code I just rearranged it

**Pull request merged**